Internal Exception: org.postgresql.util.PSQLException: ERROR: deadlock detected

Sometimes, but not all the time, when I run the API test suite on my laptop, some of the API tests fail with a 500 response and when I look at server.log I see errors like this:

`Internal Exception: org.postgresql.util.PSQLException: ERROR: deadlock detected`

Here's an example of a failing test but the tests vary:

[ERROR]   FilesIT.testReplaceFileBadJson:898 Expected status code <200> doesn't match actual status code <500>.

Here are a couple more examples from last week, both from the same run:

[ERROR] FileMetadataIT.tearDownDataverse:96 Expected status code <200> doesn't match actual status code <500>.

[ERROR] FilesIT.testReplaceFileBadJson:898 Expected status code <200> doesn't match actual status code <500>.

In tech hours today we discussed how these local failures are probably related to failures we've seen on https://jenkins.dataverse.org

Here's my server.log (renamed so GitHub will take it) from today's run: 
[server.log.txt](https://github.com/IQSS/dataverse/files/4548577/server.log.txt)

--- a/src/main/java/edu/harvard/iq/dataverse/DataverseServiceBean.java
+++ b/src/main/java/edu/harvard/iq/dataverse/DataverseServiceBean.java
@@ -16,6 +16,7 @@ import edu.harvard.iq.dataverse.batch.util.LoggingUtil;
 import edu.harvard.iq.dataverse.dataaccess.ImageThumbConverter;
 import edu.harvard.iq.dataverse.engine.command.DataverseRequest;
 import edu.harvard.iq.dataverse.search.IndexServiceBean;
+import edu.harvard.iq.dataverse.search.SolrIndexServiceBean;
 import edu.harvard.iq.dataverse.search.SolrSearchResult;
 import edu.harvard.iq.dataverse.util.SystemConfig;
 import java.io.File;
@@ -35,14 +36,11 @@ import javax.inject.Inject;
 import javax.inject.Named;
 import javax.json.Json;
 import javax.json.JsonArrayBuilder;
-import javax.json.JsonObject;
-import javax.json.JsonObjectBuilder;
 import javax.persistence.EntityManager;
 import javax.persistence.NoResultException;
 import javax.persistence.NonUniqueResultException;
 import javax.persistence.PersistenceContext;
 import javax.persistence.TypedQuery;
-import javax.ws.rs.core.Response;
 import org.apache.solr.client.solrj.SolrServerException;
 
 /**
@@ -56,6 +54,9 @@ public class DataverseServiceBean implements java.io.Serializable {
     private static final Logger logger = Logger.getLogger(DataverseServiceBean.class.getCanonicalName());
     @EJB
     IndexServiceBean indexService;
+    
+    @EJB
+    SolrIndexServiceBean solrIndexService; 
 
     @EJB
     AuthenticationServiceBean authService;
@@ -91,21 +92,33 @@ public class DataverseServiceBean implements java.io.Serializable {
        
         dataverse.setModificationTime(new Timestamp(new Date().getTime()));
         Dataverse savedDataverse = em.merge(dataverse);
+        return savedDataverse;
+    }
+    
+    public boolean index(Dataverse dataverse) {
+        return index(dataverse, false);
+
+    }
+        
+    public boolean index(Dataverse dataverse, boolean indexPermissions) {    
         /**
          * @todo check the result to see if indexing was successful or not
          * added logging of exceptions 
          */
         try {
-            Future<String> indexingResult = indexService.indexDataverse(savedDataverse);
+            indexService.indexDataverse(dataverse);
+            if (indexPermissions) {
+                solrIndexService.indexPermissionsOnSelfAndChildren(dataverse);
+            }
         } catch (IOException | SolrServerException e) {
-            String failureLogText = "Post-save indexing failed. You can kickoff a re-index of this dataverse with: \r\n curl http://localhost:8080/api/admin/index/dataverses/" + savedDataverse.getId().toString();
+            String failureLogText = "Post-save indexing failed. You can kickoff a re-index of this dataverse with: \r\n curl http://localhost:8080/api/admin/index/dataverses/" + dataverse.getId().toString();
             failureLogText += "\r\n" + e.getLocalizedMessage();
-            LoggingUtil.writeOnSuccessFailureLog(null, failureLogText, savedDataverse);
+            LoggingUtil.writeOnSuccessFailureLog(null, failureLogText, dataverse);
+            return false;
         }
 
-//        logger.log(Level.INFO, "during dataverse save, indexing result was: {0}", indexingResult);
-        return savedDataverse;
-    }
+        return true;
+    }    
 
     public Dataverse find(Object pk) {
         return em.find(Dataverse.class, pk);

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java
@@ -123,6 +123,7 @@ public abstract class AbstractCreateDatasetCommand extends AbstractDatasetComman
         // Now we need the acutal dataset id, so we can start indexing.
         ctxt.em().flush();
         
+        // TODO: this needs to be moved in to an onSuccess method; not adding to this PR as its out of scope
         // TODO: switch to asynchronous version when JPA sync works
         // ctxt.index().asyncIndexDataset(theDataset.getId(), true); 
         try{
@@ -132,9 +133,7 @@ public abstract class AbstractCreateDatasetCommand extends AbstractDatasetComman
             failureLogText += "\r\n" + e.getLocalizedMessage();
             LoggingUtil.writeOnSuccessFailureLog(null, failureLogText, theDataset);
         }
-         
-        ctxt.solrIndex().indexPermissionsOnSelfAndChildren(theDataset.getId());
-        
+                 
         return theDataset;
     }
 

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDataverseCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDataverseCommand.java
@@ -105,7 +105,7 @@ public class CreateDataverseCommand extends AbstractCommand<Dataverse> {
         DataverseRole adminRole = ctxt.roles().findBuiltinRoleByAlias(DataverseRole.ADMIN);
         String privateUrlToken = null;
 
-        ctxt.roles().save(new RoleAssignment(adminRole, getRequest().getUser(), managedDv, privateUrlToken));
+        ctxt.roles().save(new RoleAssignment(adminRole, getRequest().getUser(), managedDv, privateUrlToken),false);
         // Add additional role assignments if inheritance is set
         boolean inheritAllRoles = false;
         String rolesString = ctxt.settings().getValueForKey(SettingsServiceBean.Key.InheritParentRoleAssignments, "");
@@ -130,13 +130,13 @@ public class CreateDataverseCommand extends AbstractCommand<Dataverse> {
                             if (identifier.startsWith(AuthenticatedUser.IDENTIFIER_PREFIX)) {
                                 identifier = identifier.substring(AuthenticatedUser.IDENTIFIER_PREFIX.length());
                                 ctxt.roles().save(new RoleAssignment(role.getRole(),
-                                        ctxt.authentication().getAuthenticatedUser(identifier), managedDv, privateUrlToken));
+                                        ctxt.authentication().getAuthenticatedUser(identifier), managedDv, privateUrlToken),false);
                             } else if (identifier.startsWith(Group.IDENTIFIER_PREFIX)) {
                                 identifier = identifier.substring(Group.IDENTIFIER_PREFIX.length());
                                 Group roleGroup = ctxt.groups().getGroup(identifier);
                                 if (roleGroup != null) {
                                     ctxt.roles().save(new RoleAssignment(role.getRole(),
-                                            roleGroup, managedDv, privateUrlToken));
+                                            roleGroup, managedDv, privateUrlToken),false);
                                 }
                             }
                         }
@@ -146,6 +146,7 @@ public class CreateDataverseCommand extends AbstractCommand<Dataverse> {
         }
 
         managedDv.setPermissionModificationTime(new Timestamp(new Date().getTime()));
+        // TODO: save is called here and above; we likely don't need both
         managedDv = ctxt.dataverses().save(managedDv);
 
   //      ctxt.index().indexDataverse(managedDv);
@@ -168,17 +169,8 @@ public class CreateDataverseCommand extends AbstractCommand<Dataverse> {
     }
     
     @Override
-    public boolean onSuccess(CommandContext ctxt, Object r) {
-        try{
-            ctxt.index().indexDataverse((Dataverse) r);
-        } catch (IOException | SolrServerException e){           
-            Dataverse dv = (Dataverse) r;            
-            String failureLogText = "Indexing failed. You can kickoff a re-index of this dataverse with: \r\n curl http://localhost:8080/api/admin/index/dataverses/" + dv.getId().toString();
-            failureLogText += "\r\n" + e.getLocalizedMessage();
-            LoggingUtil.writeOnSuccessFailureLog(this, failureLogText,  dv);
-            return false;
-        }
-        return true;
+    public boolean onSuccess(CommandContext ctxt, Object r) {  
+        return ctxt.dataverses().index((Dataverse) r);
     }
 
 }

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/MoveDataverseCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/MoveDataverseCommand.java
@@ -295,6 +295,7 @@ public class MoveDataverseCommand extends AbstractVoidCommand {
         long moveDvEnd = System.currentTimeMillis();
         logger.info("Dataverse move took " + (moveDvEnd - moveDvStart) + " milliseconds");
         
+	//TODO: indexing should be moved to an on Success method
         ctxt.indexBatch().indexDataverseRecursively(moved);
         
         //REindex datasets linked to moved dv

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/PublishDataverseCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/PublishDataverseCommand.java
@@ -61,9 +61,7 @@ public class PublishDataverseCommand extends AbstractCommand<Dataverse> {
     
     @Override
     public boolean onSuccess(CommandContext ctxt, Object r) {
-        Dataverse ret = (Dataverse) r;
-        ctxt.solrIndex().indexPermissionsOnSelfAndChildren(ret.getId());
-        return true;
+        return ctxt.dataverses().index((Dataverse) r,true);
     }
 
 }

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/RenameDataverseCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/RenameDataverseCommand.java
@@ -35,4 +35,9 @@ public class RenameDataverseCommand extends AbstractCommand<Dataverse>{
 		return ctxt.dataverses().save(renamed);
 	}
 	
+        @Override
+        public boolean onSuccess(CommandContext ctxt, Object r) {
+            return ctxt.dataverses().index((Dataverse) r);
+        }  	
+	
 }

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDataverseCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDataverseCommand.java
@@ -90,20 +90,19 @@ public class UpdateDataverseCommand extends AbstractCommand<Dataverse> {
         
     @Override
     public boolean onSuccess(CommandContext ctxt, Object r) {
-        Dataverse result = (Dataverse) r;
-
-        List<Dataset> datasets = ctxt.datasets().findByOwnerId(result.getId());
+        
+        // first kick of async index of datasets
+        // TODO: is this actually needed? Is there a better way to handle
         try {
-            Future<String> indResponse = ctxt.index().indexDataverse(result);
+            Dataverse result = (Dataverse) r;
+            List<Dataset> datasets = ctxt.datasets().findByOwnerId(result.getId());
             ctxt.index().asyncIndexDatasetList(datasets, true);
         } catch (IOException | SolrServerException e) {
-            String failureLogText = "Indexing failed for Updated Dataverse. You can kickoff a re-index of this datavese with: \r\n curl http://localhost:8080/api/admin/index/datasets/" + result.getId().toString();
-            failureLogText += "\r\n" + e.getLocalizedMessage();
-            LoggingUtil.writeOnSuccessFailureLog(this, failureLogText, result);
-            return false;
+            // these datasets are being indexed asynchrounously, so not sure how to handle errors here
         }
-        return true;
-    }
+        
+        return ctxt.dataverses().index((Dataverse) r);
+    }  
 
 }
 

--- a/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdatePermissionRootCommand.java
+++ b/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdatePermissionRootCommand.java
@@ -38,6 +38,13 @@ public class UpdatePermissionRootCommand extends AbstractCommand<Dataverse> {
 	}
 
     @Override
+    public boolean onSuccess(CommandContext ctxt, Object r) {  
+        return ctxt.dataverses().index((Dataverse) r,true);
+    }        
+
+
+    //TODO: Review this as this will never be an instance of Dataset, will it?
+    @Override
     public Map<String, Set<Permission>> getRequiredPermissions() {
         // for data file check permission on owning dataset
         return Collections.singletonMap("",

--- a/src/test/java/edu/harvard/iq/dataverse/api/DataversesIT.java
+++ b/src/test/java/edu/harvard/iq/dataverse/api/DataversesIT.java
@@ -2,7 +2,6 @@ package edu.harvard.iq.dataverse.api;
 
 import com.jayway.restassured.RestAssured;
 import static com.jayway.restassured.RestAssured.given;
-import com.jayway.restassured.path.json.JsonPath;
 import static com.jayway.restassured.path.json.JsonPath.with;
 import com.jayway.restassured.response.Response;
 import edu.harvard.iq.dataverse.Dataverse;
@@ -12,6 +11,7 @@ import java.io.FileNotFoundException;
 import java.io.FileReader;
 import java.util.Arrays;
 import java.util.List;
+import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.json.Json;
 import javax.json.JsonObject;
@@ -21,9 +21,7 @@ import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
 import javax.ws.rs.core.Response.Status;
 import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
 import static javax.ws.rs.core.Response.Status.OK;
-import static javax.ws.rs.core.Response.Status.UNAUTHORIZED;
 import static junit.framework.Assert.assertEquals;
-import static junit.framework.Assert.fail;
 import org.junit.BeforeClass;
 import org.junit.Test;
 import static org.hamcrest.CoreMatchers.equalTo;
@@ -362,11 +360,28 @@ public class DataversesIT {
         moveResponse.prettyPrint();
         moveResponse.then().assertThat().statusCode(OK.getStatusCode());
         
-        Response search = UtilIT.search("id:dataverse_" + dataverseId + "&subtree=" + dataverseAlias2, apiToken);
-        search.prettyPrint();
-        search.then().assertThat()
-                .body("data.total_count", equalTo(1))
-                .statusCode(200);
+        // because indexing happens asynchronously, we'll wait first, and then retry a few times, before failing
+        int numberofAttempts = 0;
+        boolean checkIndex = true;
+        while (checkIndex) {
+            try {   
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException ex) {
+                    }                
+                Response search = UtilIT.search("id:dataverse_" + dataverseId + "&subtree=" + dataverseAlias2, apiToken);
+                search.prettyPrint();
+                search.then().assertThat()
+                        .body("data.total_count", equalTo(1))
+                        .statusCode(200);
+                checkIndex = false;
+            } catch (AssertionError ae) {
+                if (numberofAttempts++ > 5) {
+                    throw ae;
+                }
+            }
+        }
+
     }
     
     @Test

--- a/src/test/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDataverseCommandTest.java
+++ b/src/test/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDataverseCommandTest.java
@@ -73,6 +73,12 @@ public class CreateDataverseCommandTest {
             return dataverse;
         }
         
+        @Override
+        public boolean index(Dataverse dataverse) {
+            indexCalled=true;
+            return true;
+        }        
+        
     };
     
     DataverseRoleServiceBean roles = new DataverseRoleServiceBean(){
@@ -99,6 +105,11 @@ public class CreateDataverseCommandTest {
             assignments.add(assignment);
             return assignment;
         }
+        
+        @Override
+        public RoleAssignment save(RoleAssignment assignment, boolean index) {
+            return save (assignment);
+        }        
 
         @Override
         public List<RoleAssignment> directRoleAssignments(DvObject dvo) {

--- a/src/test/java/edu/harvard/iq/dataverse/engine/command/impl/UpdatePermissionRootCommandTest.java
+++ b/src/test/java/edu/harvard/iq/dataverse/engine/command/impl/UpdatePermissionRootCommandTest.java
@@ -30,6 +30,11 @@ public class UpdatePermissionRootCommandTest {
                 serviceBeanCalled = true;
                 return dv;
             }
+            @Override
+            public boolean index(Dataverse dataverse, boolean indexPermisions) {
+                    // no-op. The superclass accesses databases which we don't have.
+                    return true;
+            }            
         };
         testCommandContext = new TestCommandContext() {
             @Override

